### ISSUE TITLE
Character '"' is not valid at position 20 in 'PartitionKey='"A'aaa"_bbbb3',RowKey='"A'aaa"_bbbb3''.

### ISSUE THREAD

**Issue by 646826** [Azure/azure-sdk-for-python#20301]
- **Package Name**: azure-data-tables
- **Package Version**: 12.1.0
- **Operating System**: Linux and Windows
- **Python Version**: 3.8

**Describe the bug**
```
TableTransactionError: 0:Syntax error: character '"' is not valid at position 20 in 'PartitionKey='"A'aaa"_bbbb3',RowKey='"A'aaa"_bbbb3''.
RequestId:8452bf63-8002-0099-18eb-9265f1000000
Time:2021-08-16T22:08:14.5833819Z
ErrorCode:InvalidInput
---------------------------------------------------------------------------
TableTransactionError                     Traceback (most recent call last)
/Untitled-1.py in 
     15 azure_table_client = TableServiceClient(endpoint=f"https://{account_name}.table.core.windows.net", credential=credential)
     16 azure_table_service = azure_table_client .create_table_if_not_exists(to_table_name)
---> 17 azure_table_service .submit_transaction(batch)

/usr/local/lib/python3.8/site-packages/azure/data/tables/_table_client.py in submit_transaction(self, operations, **kwargs)
    729             except AttributeError:
    730                 raise ValueError("Unrecognized operation: {}".format(operation[0]))
--> 731         return self._batch_send(*batched_requests.requests, **kwargs)  # type: ignore

/usr/local/lib/python3.8/site-packages/azure/data/tables/_base_client.py in _batch_send(self, *reqs, **kwargs)
    301                     error_message="The transaction request was too large",
    302                     error_type=RequestTooLargeError)
--> 303             raise _decode_error(
    304                 response=error_parts[0],
    305                 error_type=TableTransactionError

TableTransactionError: 0:Syntax error: character '"' is not valid at position 20 in 'PartitionKey='"A'aaa"_bbbb3',RowKey='"A'aaa"_bbbb3''.
RequestId:8452bf63-8002-0099-18eb-9265f1000000
Time:2021-08-16T22:08:14.5833819Z
ErrorCode:InvalidInput
```

**To Reproduce**
`pip install azure-data-tables`
```
from azure.core.credentials import AzureNamedKeyCredential
from azure.data.tables import TableServiceClient
from azure.data.tables import TableEntity

account_name=""
account_key=""
to_table_name=""

batch = []
g = TableEntity({"PartitionKey":"\"A'aaa\"_bbbb3", "RowKey":"\"A'aaa\"_bbbb3", "test":"\"A'aaa\"_bbbb3"})
batch.append(("upsert", TableEntity(g)))

credential = AzureNamedKeyCredential(account_name, account_key)
azure_table_client = TableServiceClient(endpoint=f"https://{account_name}.table.core.windows.net", credential=credential)
azure_table_service = azure_table_client.create_table_if_not_exists(to_table_name)
azure_table_service.submit_transaction(batch)
```

**Expected behavior**
- I want to create this PartitionKey: "A'aaa\"_bbbb3  
- I  want to use this value "A'aaa\"_bbbb3 for rows  

**Screenshots**
![image](https://user-images.githubusercontent.com/1565611/129635848-f83ff148-572f-40aa-b310-5c72421b4fd5.png)

**Additional context**
I was able to use these values via:
> Microsoft Azure Storage Explorer version 1.20.1   
![image](https://user-images.githubusercontent.com/1565611/129636024-2f06d53d-b4a0-43e8-92ca-b8df134e559b.png)  
>and Storage Explorer (preview) at https://portal.azure.com/  
![image](https://user-images.githubusercontent.com/1565611/129637250-d83c0afd-9f5b-4137-a35a-e6a39c9bdca3.png)  

**#Update from me**
I tried loading data through query_entities, I got this result:
```
{'PartitionKey': '"A\'aaa"_bbbb2',
 'RowKey': '"A\'aaa"_bbbb2',
 'test': '"A\'aaa"_bbbb2'}
```
![image](https://user-images.githubusercontent.com/1565611/129638157-85251d13-2323-4e23-bcca-5bcc5686f2e5.png)

After that, I tried to use the response to send this data to the server but I got the same error.

**Comment by annatisch** [comment]
Thanks for the report @646826!
Looking at [the service documentation](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-the-table-service-data-model#system-properties) - the partition key and row key values here _should_ be valid. So this seems like a bug in the SDK.

Have you tried the same `upsert` operation as a single request rather than a transaction (e.g. `client.upsert_entity(...)`)?
Additionally - looking at your code sample - you shouldn't need to wrap your input entity in `TableEntity`, the following should also be valid:
```python
batch = []
g = {"PartitionKey":"\"A'aaa\"_bbbb3", "RowKey":"\"A'aaa\"_bbbb3", "test":"\"A'aaa\"_bbbb3"}
batch.append(("upsert", g))
```

I will try to repro this in a test asap.

**Comment by 646826** [comment]
> Thanks for the report @646826!
> Looking at [the service documentation](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-the-table-service-data-model#system-properties) - the partition key and row key values here _should_ be valid. So this seems like a bug in the SDK.
> 
> Have you tried the same `upsert` operation as a single request rather than a transaction (e.g. `client.upsert_entity(...)`)?
> Additionally - looking at your code sample - you shouldn't need to wrap your input entity in `TableEntity`, the following should also be valid:
> 
> ```python
> batch = []
> g = {"PartitionKey":"\"A'aaa\"_bbbb3", "RowKey":"\"A'aaa\"_bbbb3", "test":"\"A'aaa\"_bbbb3"}
> batch.append(("upsert", g))
> ```
> 
> I will try to repro this in a test asap.

I spent a lot of time trying to write/update this data to a table. And I was not completely sure where the problem was, so I tried all possible options. I have added some additional information to the bug, please see above.

**Comment by annatisch** [comment]
Thanks for your patience - I was able to successfully repro this locally. Investigating the cause now :)

**Comment by annatisch** [comment]
Okay - I have found the cause for this. It's the single quote character `'`, the service requires that this be escaped via duplication.
For example, the entity you have in your example:
```python
g = {"PartitionKey":"\"A'aaa\"_bbbb3", "RowKey":"\"A'aaa\"_bbbb3", "test":"\"A'aaa\"_bbbb3"}
```
When sent over the wire, would need to be formatted like this:
```python
g = {"PartitionKey":"\"A''aaa\"_bbbb3", "RowKey":"\"A''aaa\"_bbbb3", "test":"\"A'aaa\"_bbbb3"}
```
This is working correctly for deserialization, but not serialization.
I will PR a fix this week to ship in the September release.

**Comment by YalinLi0312** [comment]
Fixed.

**Comment by annatisch** [comment]
Unfortunately we have moved this release to the beginning of October. We are still waiting on cross-language consensus on whether this should be considered a breaking change, and if not, how an 'opt-out' strategy will be adopted for each SDK.

Current opt-out strategy for Python is implemented in this PR, pending further design discussions:
https://github.com/Azure/azure-sdk-for-python/pull/20603

I have reopened this issue for now.
